### ISSUE TITLE
Undetected `#Bottom` nodes  leads to infinite branching in combination with the `extract_branches` mechanism

### ISSUE THREAD

**Issue by lisandrasilva** [runtimeverification/kontrol#127]
A proof had the following unfeasible branch: 
```
...
#And { maxUInt8 <Int bool2Word ( maxUInt128 <Int VV0_x_114b9705:Int ) <<Int 7 #Equals true } 
...
```
After writing the following lemma to discard the branch:
```
rule A <Int bool2Word ( _ ) <<Int B => false
      requires A >=Int 1 <<Int B [simplification, concrete( A , B )]
```
The proof kept branching infinitely on the same condition: 
```
true #Equals 1 <<Word bool2Word ( maxUInt128 <Int VV0_x_114b9705:Int ) <<Int 7 <=Int VV0_x_114b9705:Int }', '{ true #Equals ( notBool 1 <<Word bool2Word ( maxUInt128 <Int VV0_x_114b9705:Int ) <<Int 7 <=Int VV0_x_114b9705:Int ) }
```

The proof that allows to reproduce the bug can be found [here](https://github.com/runtimeverification/kontrol-solady/blob/358a53c3b226f2c0bd24614d294213f17f2c81bb/test/FixedPointMathLib.k.sol#L24). One can reproduce the bug by cloning this [branch](https://github.com/runtimeverification/kontrol-solady/tree/log2-simpler) and run the script `run-kevm.sh`.

After some investigation @PetarMax found out that it was due to a bug in the `extract_branches` mechanism. 

----------------------------------
~[@geo2a] Blocked on https://github.com/runtimeverification/evm-semantics/issues/2151~
~[@geo2a] Pending discussion on making `<<Word` look like an uninterpreted function to Z3, i.e. adding `smtlib(shlWord)` to its definition in `evm-semantics`.~
[@geo2a] Blocked on https://github.com/runtimeverification/evm-semantics/issues/2151 --- adding `smtlib` attribute to operations on `Word`.

**Comment by PetarMax** [comment]
The way I discovered it by commenting out the `extract_branches` logic in KEVM to just return `[]` and re-running the proof.

**Comment by PetarMax** [comment]
I've investigated further. I instrumented `pyk` to get some more prints and ran 
```
kontrol build --verbose --require solady-lemmas.k --module-import FixedPointMathLibVerification:SOLADY-LEMMAS --rekompile --regen --foundry-project-root ..
``` 
and 
```
kontrol prove --max-depth 3000 --max-iterations 10 --smt-timeout 1000 --verbose --no-break-on-calls --test FixedPointMathLibVerification.testMyLog2 --foundry-project-root ..
```
from the `test` folder of the branch @lisandrasilva provided above. The problem that I've spotted is that the `simplify` endpoint is not able to discharge the following:
```
( #Not ( { VV0_x_114b9705:Int #Equals 0 } )
#And ( { 0 <=Int CALLER_ID:Int #Equals true }
#And ( { 0 <=Int ORIGIN_ID:Int #Equals true }
#And ( { 0 <=Int NUMBER_CELL:Int #Equals true }
#And ( { 0 <=Int VV0_x_114b9705:Int #Equals true }
#And ( { CALLER_ID:Int <Int pow160 #Equals true }
#And ( { ORIGIN_ID:Int <Int pow160 #Equals true }
#And ( { NUMBER_CELL:Int <=Int maxSInt256 #Equals true }
#And ( { VV0_x_114b9705:Int <Int pow256 #Equals true }
#And ( #Not ( { 0 #Equals bool2Word ( maxUInt128 <Int VV0_x_114b9705:Int ) <<Int 7 } )
#And ( { true #Equals ( notBool 1 <<Word bool2Word ( maxUInt128 <Int VV0_x_114b9705:Int ) <<Int 7 <=Int VV0_x_114b9705:Int ) }
#And { true #Equals 1 <<Word bool2Word ( maxUInt128 <Int VV0_x_114b9705:Int ) <<Int 7 <=Int VV0_x_114b9705:Int } ) ) ) ) ) ) ) ) ) ) )
```
into `#Bottom` (the last two lines are a contradiction), returning instead
```
( ( { true #Equals CALLER_ID:Int <Int pow160 }
#And ( { true #Equals ORIGIN_ID:Int <Int pow160 }
#And { true #Equals VV0_x_114b9705:Int <Int 1 <<Word bool2Word ( maxUInt128 <Int VV0_x_114b9705:Int ) <<Int 7 } ) )
#And ( { true #Equals VV0_x_114b9705:Int <Int pow256 }
#And ( ( { true #Equals 1 <<Word bool2Word ( maxUInt128 <Int VV0_x_114b9705:Int ) <<Int 7 <=Int VV0_x_114b9705:Int }
#And ( { true #Equals 0 <=Int CALLER_ID:Int }
#And { true #Equals 0 <=Int NUMBER_CELL:Int } ) )
#And ( { true #Equals 0 <=Int ORIGIN_ID:Int }
#And ( { true #Equals 0 <=Int VV0_x_114b9705:Int }
#And ( { true #Equals NUMBER_CELL:Int <=Int maxSInt256 }
#And ( #Not ( { 0 #Equals bool2Word ( maxUInt128 <Int VV0_x_114b9705:Int ) <<Int 7 } )
#And #Not ( { VV0_x_114b9705:Int #Equals 0 } ) ) ) ) ) ) ) )
```
with the third and fifth constraint being in contradiction. 

This happens regardless of whether or not the booster is used. I am attaching the bug report, the last several calls to the RPC server should be relevant. I wonder if the SMT solver is being/should be called during this process or if any of the operations present make it impossible for it to be called.

Bringing in @goodlyrottenapple and @geo2a.

**Comment by PetarMax** [comment]
...and here's the bug report: [report.tar.gz](https://github.com/runtimeverification/kontrol/files/13189310/report.tar.gz).

**Comment by geo2a** [comment]
Just to config, @PetarMax and @lisandrasilva which versions of `kontrol`, `kore-rpc` and `kore-rpc-booster` are you using? Am I right that you're using Nix overrides to get the latest `kore-rpc-booster`? It seems like maybe in a little bit of a pickle, because the recent changes to the presentation of `#And` and `#Or` across several codebases may be backfiring, i.e. your `kore-rpc`/`kore-rpc-booster` gives something that your `pyk` and `kontrol` don't understand quite right. I'll try consolidating my versions to a good state tomorrow and re-running the proof from scratch.

**Comment by geo2a** [comment]
I can confirm that we have a three-way branch, as above. I've managed to isolate the request that makes `kore-rpc`  branch with 3 "next-states" at depth 0. I've nested with commits `811e94fbcaef8550a6a709a9ce793f7166457911` and `2a99e06e67fb408a1219459978177fbb7843dee0`.

Attached are the definition, request which brances at depth 0 and the pretty-printed response:
[request_three_way_branch.json](https://github.com/runtimeverification/kontrol/files/13216548/request_three_way_branch.json)
[definition.kore.txt](https://github.com/runtimeverification/kontrol/files/13216543/definition.kore.txt)
[response_three_way_branch_kore-rpc-dev.pretty.txt](https://github.com/runtimeverification/kontrol/files/13216550/response_three_way_branch_kore-rpc-dev.pretty.txt)

**Comment by geo2a** [comment]
I'm tracking the three-way branch issue as https://github.com/runtimeverification/haskell-backend/issues/3690

There's potentially another issue, i.e. the one originally discussed by Petar, where we cannot discharge a set of contradictory constraints. I am still to isolate it and put into a separate ticket.

**Comment by geo2a** [comment]
The `<<Word` function does not have an `smtlib` attribute. Adding the attribute manually in `definition.kore`:
```
  symbol Lbl'Unds-LT--LT-'Word'UndsUnds'EVM-TYPES'Unds'Int'Unds'Int'Unds'Int{}(SortInt{}, SortInt{}) : SortInt{} [smtlib{}("shlWord"), format{}("%1 %c<<Word%r %2"), function{}(), functional{}(), left{}(), priorities{}(), right{}(), terminals{}("010"), total{}()]
```
makes the constraints the contradictory constraints

```
#And{ true #Equals VV0_x_114b9705:Int <Int 1 <<Word bool2Word ( maxUInt128 <Int VV0_x_114b9705:Int ) <<Int 7 }
#And{ true #Equals 1 <<Word bool2Word ( maxUInt128 <Int VV0_x_114b9705:Int ) <<Int 7 <=Int VV0_x_114b9705:Int }
```

simplify to `#Bottom` and the imaginary third branch goes away, leaving only the two actual branches in the KCFG. Once of the branches is immediately discharged, and execution continues along the second one.

@lisandrasilva @ehildenb @PetarMax what do you think? Should we add `smtlib` attributes to all functions on `Word`, like we have them on functions on `Int` in [domains.md](https://kframework.org/k-distribution/include/kframework/builtin/domains/#integer-arithmetic)?

**Comment by PetarMax** [comment]
I think that we might need a bit of external advice, since we didn't design the EVM semantics - @ehildenb, what is the reason for some of the functions having the `smtlib` attribute and some not?

**Comment by geo2a** [comment]
It seems like the next step is to add `smtlabel` attributes to `Word`-related functions. This is tracked by https://github.com/runtimeverification/evm-semantics/issues/2151. I'll modify the issue description to reflect that too.

**Comment by yale-vinson** [comment]
@geo2a Is there any progress on getting this Issue closed out?

**Comment by Baltoli** [comment]
Needs a new reproduction (@geo2a); it's been a while since the original issue was reported and there have been a number of different changes to the backend. Since the issue was reported, the booster has gained an SMT solver so we should see if it can discharge the conditions itself without reverting.

**Comment by palinatolmach** [comment]
Closing as it doesn't seem to be an issue anymore; will reopen if happens again.
